File: Modules/SekaiClient/manager.py
import asyncio
import aiofiles
import ujson as json
from pathlib import Path
from typing import Dict, Tuple, List, Union, Optional
from tenacity import retry, stop_after_attempt, wait_fixed, retry_if_exception_type

from ..logger import AsyncLogger
from .client import SekaiClient
from .helper import SekaiCookieHelper, SekaiVersionHelper
from .exceptions import UpgradeRequiredError, UnderMaintenanceError, CookieExpiredError
from .model import SekaiServerInfo, SekaiAccountCP, SekaiAccountNuverse, SekaiServerRegion

logger = AsyncLogger(__name__, level="DEBUG")


class SekaiClientManager:
    def __init__(
        self,
        server_info: SekaiServerInfo,
        accounts_dir: Union[Path, str],
        version_file_path: Union[Path, str],
        proxies: Optional[List] = None,
    ) -> None:
        # Server configs
        self.server: SekaiServerRegion = SekaiServerRegion(server_info.server)
        self.server_info: SekaiServerInfo = server_info
        self.version_helper: SekaiVersionHelper = SekaiVersionHelper(version_file_path)
        self.cookie_helper: Union[SekaiCookieHelper, None] = (
            SekaiCookieHelper() if self.server == SekaiServerRegion.JP else None
        )
        # Account configs
        self.accounts_dir: Union[Path, str] = accounts_dir
        self.clients: List[SekaiClient] = []
        self.client_no: int = 0
        # Proxies config
        self.proxies = proxies

    # Generate an account pool
    async def _parse_accounts(self) -> List[Union[SekaiAccountCP, SekaiAccountNuverse]]:
        accounts = []
        for json_file in Path(self.accounts_dir).rglob("*.json"):
            async with aiofiles.open(json_file, mode="r", encoding="utf-8") as f:
                try:
                    data = json.loads(await f.read())
                except json.JSONDecodeError as e:
                    await logger.warning(f"Error decoding JSON in file {json_file}: {e}")
                    continue
                if isinstance(data, dict):
                    if self.server in [SekaiServerRegion.JP, SekaiServerRegion.EN]:
                        account = SekaiAccountCP(**data)
                    else:
                        account = SekaiAccountNuverse(**data)
                    accounts.append(account)
                elif isinstance(data, list):
                    for _account in data:
                        if self.server in [SekaiServerRegion.JP, SekaiServerRegion.EN]:
                            account = SekaiAccountCP(**_account)
                        else:
                            account = SekaiAccountNuverse(**_account)
                        accounts.append(account)
                else:
                    await logger.warning(f"Unexpected data type in file {json_file}: {type(data)}")
                    continue

        return accounts

    # Re-parse cookies
    async def _parse_cookies(self) -> None:
        if self.server == SekaiServerRegion.JP:
            tasks = [client.parse_cookies() for client in self.clients]
            await asyncio.gather(*tasks)

    # Re-parse version data
    async def _parse_version(self) -> None:
        tasks = [client.parse_version() for client in self.clients]
        await asyncio.gather(*tasks)

    # Init manager
    async def init(self) -> None:
        await logger.start()
        # Create clients list
        _accounts = await self._parse_accounts()
        self.clients.extend(
            [
                SekaiClient(
                    self.server_info,
                    _account,
                    self.server_info.aes_key,
                    self.server_info.aes_iv,
                    self.cookie_helper,
                    self.version_helper,
                    self.proxies,
                )
                for _account in _accounts
            ]
        )
        # Init clients
        client_init_tasks = [client.init() for client in self.clients]
        await asyncio.gather(*client_init_tasks)

        # Login
        try:
            login_tasks = [client.login() for client in self.clients]
            await asyncio.gather(*login_tasks)
        except Exception as e:
            await logger.error(f"Error while initializing Sekai client: {e}")

    # Get a client
    async def get_client(self) -> Optional[SekaiClient]:
        if self.client_no == len(self.clients):
            self.client_no = 0
            return self.clients[self.client_no]
        else:
            self.client_no += 1
            return self.clients[self.client_no - 1]

    # Get login data
    async def get_login_data(self) -> Optional[Dict]:
        client = await self.get_client()
        if not client.lock.locked():
            async with client.lock:
                return await client.login()
        return None

    # Download master data
    async def download_master(self) -> Optional[Dict]:
        client = await self.get_client()
        if not client.lock.locked():
            async with client.lock:
                return await client.download_master()
        return None

    # Call game API
    @retry(
        stop=stop_after_attempt(4),
        wait=wait_fixed(1),
        retry=retry_if_exception_type((UpgradeRequiredError, CookieExpiredError)),
    )
    async def api_get(self, path: str, params: Optional[Dict] = None) -> Optional[Tuple[Dict, int]]:
        client = await self.get_client()
        if not client.lock.locked():
            async with client.lock:
                try:
                    return await client.get(path, params=params)
                except CookieExpiredError:
                    await logger.warning(f"{self.server.value.upper()} Server cookies expired, re-parsing...")
                    await self._parse_cookies()
                except UpgradeRequiredError:
                    await logger.warning(f"{self.server.value.upper()} Server upgrade required, re-parsing...")
                    await self._parse_version()
                except UnderMaintenanceError:
                    await logger.warning(f"{self.server.value.upper()} Server is under maintenance.")
                    error_response = {
                        "result": "failed",
                        "message": f"{self.server.value.upper()} Game server is under maintenance.",
                    }
                    return error_response, 503
                except Exception as e:
                    await logger.warning(f"Failed to call {self.server.value.upper()} Server API: {repr(e)}")
                    error_response = {"result": "failed", "message": repr(e)}
                    return error_response, 500
        else:
            error_response = {"result": "failed", "message": "No client is available, please try again later."}
            return error_response, 500

    async def image_get(self, path: str) -> Optional[Tuple[Union[bytes, str], int]]:
        client = await self.get_client()
        if not client.lock.locked():
            return await client.get_image(path)
        return None

    async def shutdown(self) -> None:
        tasks = [client.close() for client in self.clients]
        await asyncio.gather(*tasks)
        await logger.stop()

File: client/manager.go
package client

import (
	"context"
	"fmt"
	"haruki-sekai-api/utils"
	"haruki-sekai-api/utils/git"
	"haruki-sekai-api/utils/logger"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/bytedance/sonic"
)

type SekaiClientManager struct {
	Server              utils.HarukiSekaiServerRegion
	ServerConfig        utils.HarukiSekaiServerConfig
	VersionHelper       *SekaiVersionHelper
	CookieHelper        *SekaiCookieHelper
	Clients             []*SekaiClient
	AssetUpdaterServers []*utils.HarukiAssetUpdaterInfo
	Git                 *git.HarukiGitUpdater
	ClientNo            int
	Proxy               string
	Logger              *logger.Logger
}

func NewSekaiClientManager(server utils.HarukiSekaiServerRegion, serverConfig utils.HarukiSekaiServerConfig, assetUpdaterServers []*utils.HarukiAssetUpdaterInfo, git *git.HarukiGitUpdater, proxy string, jpSekaiCookieURL string) *SekaiClientManager {
	mgr := &SekaiClientManager{
		Server:              server,
		ServerConfig:        serverConfig,
		VersionHelper:       &SekaiVersionHelper{versionFilePath: serverConfig.VersionPath},
		Proxy:               proxy,
		AssetUpdaterServers: assetUpdaterServers,
		Git:                 git,
		Logger:              logger.NewLogger(fmt.Sprintf("SekaiClientManager%s", strings.ToUpper(string(server))), "DEBUG", nil),
	}
	if server == utils.HarukiSekaiServerRegionJP {
		mgr.CookieHelper = &SekaiCookieHelper{url: jpSekaiCookieURL}
	}
	return mgr
}

func (mgr *SekaiClientManager) parseAccounts() ([]SekaiAccountInterface, error) {
	var (
		accounts []SekaiAccountInterface
	)

	err := filepath.Walk(mgr.ServerConfig.AccountDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			mgr.Logger.Warnf("parseAccounts: walk error on %s: %v", path, err)
			return err
		}
		if info.IsDir() {
			return nil
		}
		if filepath.Ext(path) != ".json" {
			return nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			mgr.Logger.Warnf("parseAccounts: read error %s: %v", path, err)
			return nil
		}
		var raw any
		if err := sonic.Unmarshal(data, &raw); err != nil {
			mgr.Logger.Warnf("parseAccounts: json decode error %s: %v", path, err)
			return nil
		}

		switch v := raw.(type) {
		case map[string]any:
			if mgr.Server == utils.HarukiSekaiServerRegionJP || mgr.Server == utils.HarukiSekaiServerRegionEN {
				acc := new(SekaiAccountCP)
				b, _ := sonic.Marshal(v)
				if err := sonic.Unmarshal(b, acc); err == nil {
					accounts = append(accounts, acc)
				} else {
					mgr.Logger.Warnf("parseAccounts: CP unmarshal error %s: %v", path, err)
				}
			} else {
				acc := new(SekaiAccountNuverse)
				b, _ := sonic.Marshal(v)
				if err := sonic.Unmarshal(b, acc); err == nil {
					accounts = append(accounts, acc)
				} else {
					mgr.Logger.Warnf("parseAccounts: Nuverse unmarshal error %s: %v", path, err)
				}
			}
		case []any:
			for idx, item := range v {
				if m, ok := item.(map[string]any); ok {
					if mgr.Server == utils.HarukiSekaiServerRegionJP || mgr.Server == utils.HarukiSekaiServerRegionEN {
						acc := new(SekaiAccountCP)
						b, _ := sonic.Marshal(m)
						if err := sonic.Unmarshal(b, acc); err == nil {
							accounts = append(accounts, acc)
						} else {
							mgr.Logger.Warnf("parseAccounts: [%s][%d] CP unmarshal error: %v", path, idx, err)
						}
					} else {
						acc := new(SekaiAccountNuverse)
						b, _ := sonic.Marshal(m)
						if err := sonic.Unmarshal(b, acc); err == nil {
							accounts = append(accounts, acc)
						} else {
							mgr.Logger.Warnf("parseAccounts: [%s][%d] Nuverse unmarshal error: %v", path, idx, err)
						}
					}
				} else {
					mgr.Logger.Warnf("parseAccounts: [%s][%d] unexpected array element type: %T", path, idx, item)
				}
			}
		default:
			mgr.Logger.Warnf("parseAccounts: unexpected top-level type in %s: %T", path, v)
		}
		return nil
	})

	if len(accounts) == 0 {
		mgr.Logger.Warnf("parseAccounts: no accounts parsed from %s", mgr.ServerConfig.AccountDir)
	}

	return accounts, err
}

func (mgr *SekaiClientManager) parseCookies(ctx context.Context) error {
	if mgr.Server == utils.HarukiSekaiServerRegionJP {
		var wg sync.WaitGroup
		errChan := make(chan error, len(mgr.Clients))
		for _, client := range mgr.Clients {
			wg.Add(1)
			go func(c *SekaiClient) {
				defer wg.Done()
				if err := c.ParseCookies(ctx); err != nil {
					mgr.Logger.Warnf("Error parsing cookies: %v", err)
					errChan <- err
				}
			}(client)
		}
		wg.Wait()
		close(errChan)

		for err := range errChan {
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func (mgr *SekaiClientManager) parseVersion() error {
	var wg sync.WaitGroup
	errChan := make(chan error, len(mgr.Clients))
	for _, client := range mgr.Clients {
		wg.Add(1)
		go func(c *SekaiClient) {
			defer wg.Done()
			if err := c.ParseVersion(); err != nil {
				mgr.Logger.Warnf("Error parsing version: %v", err)
				errChan <- err
			}
		}(client)
	}
	wg.Wait()
	close(errChan)

	for err := range errChan {
		if err != nil {
			return err
		}
	}
	return nil
}

func (mgr *SekaiClientManager) Init() error {
	mgr.Logger.Debugf("Initializing client manager...")

	accounts, err := mgr.parseAccounts()
	if err != nil {
		return err
	}

	for _, account := range accounts {
		client := NewSekaiClient(
			mgr.Server,
			mgr.ServerConfig,
			account,
			mgr.CookieHelper,
			mgr.VersionHelper,
			mgr.Proxy,
		)
		mgr.Clients = append(mgr.Clients, client)
	}

	var wg sync.WaitGroup
	initErrors := make(chan error, len(mgr.Clients))
	for _, client := range mgr.Clients {
		wg.Add(1)
		go func(c *SekaiClient) {
			defer wg.Done()
			if err := c.Init(); err != nil {
				mgr.Logger.Errorf("Error initializing client: %v", err)
				initErrors <- err
			}
		}(client)
	}
	wg.Wait()
	close(initErrors)

	for err := range initErrors {
		if err != nil {
			return err
		}
	}

	ctx := context.Background()
	loginErrors := make(chan error, len(mgr.Clients))
	for _, client := range mgr.Clients {
		wg.Add(1)
		go func(c *SekaiClient) {
			defer wg.Done()
			if _, err := c.Login(ctx); err != nil {
				mgr.Logger.Errorf("Error logging in: %v", err)
				loginErrors <- err
			}
		}(client)
	}
	wg.Wait()
	close(loginErrors)

	for err := range loginErrors {
		if err != nil {
			return err
		}
	}

	mgr.Logger.Infof("Client manager initialized successfully")
	return nil
}

func (mgr *SekaiClientManager) getClient() *SekaiClient {
	if len(mgr.Clients) == 0 {
		return nil
	}
	if mgr.ClientNo >= len(mgr.Clients) || mgr.ClientNo < 0 {
		mgr.ClientNo = 0
	}
	idx := mgr.ClientNo % len(mgr.Clients)
	c := mgr.Clients[idx]
	mgr.ClientNo = (idx + 1) % len(mgr.Clients)
	return c
}

func (mgr *SekaiClientManager) Shutdown() error {
	var wg sync.WaitGroup
	errChan := make(chan error, len(mgr.Clients))

	for _, client := range mgr.Clients {
		wg.Add(1)
		go func(c *SekaiClient) {
			defer wg.Done()
			if err := c.Close(); err != nil {
				mgr.Logger.Warnf("Error closing client: %v", err)
				errChan <- err
			}
		}(client)
	}
	wg.Wait()
	close(errChan)

	for err := range errChan {
		if err != nil {
			return err
		}
	}

	mgr.Logger.Debugf("Client manager shut down successfully")
	return nil
}

func (mgr *SekaiClientManager) GetGameAPI(ctx context.Context, path string, params map[string]any) (any, int, error) {
	if len(mgr.Clients) == 0 {
		resp := HarukiSekaiAPIFailedResponse{
			Result:  "failed",
			Status:  http.StatusInternalServerError,
			Message: "No client initialized",
		}
		return resp, http.StatusInternalServerError, nil
	}

	maxRetries := 4
	retryCount := 0
	retryDelay := time.Second

	for retryCount < maxRetries {
		client := mgr.getClient()
		if client == nil {
			resp := HarukiSekaiAPIFailedResponse{
				Result:  "failed",
				Status:  http.StatusInternalServerError,
				Message: "No client is available, please try again later.",
			}
			return resp, http.StatusInternalServerError, nil
		}

		client.Lock.Lock()

		response, err := client.Get(ctx, path, params)
		if err != nil {
			client.Lock.Unlock()
			resp := HarukiSekaiAPIFailedResponse{
				Result:  "failed",
				Status:  http.StatusInternalServerError,
				Message: err.Error(),
			}
			return resp, http.StatusInternalServerError, err
		}

		statusCode, err := ParseSekaiApiHttpStatus(response.StatusCode())
		if err != nil {
			client.Lock.Unlock()
			resp := HarukiSekaiAPIFailedResponse{
				Result:  "failed",
				Status:  response.StatusCode(),
				Message: fmt.Sprintf("Unknown status code: %d", response.StatusCode()),
			}
			return resp, response.StatusCode(), err
		}

		switch statusCode {
		case SekaiApiHttpStatusGameUpgrade:
			mgr.Logger.Warnf("%s Server upgrade required, re-parsing...", strings.ToUpper(string(mgr.Server)))
			if err := mgr.parseVersion(); err != nil {
				client.Lock.Unlock()
				resp := HarukiSekaiAPIFailedResponse{
					Result:  "failed",
					Status:  response.StatusCode(),
					Message: fmt.Sprintf("Failed to parse version: %v", err),
				}
				return resp, response.StatusCode(), err
			}
			retryCount++
			time.Sleep(retryDelay)
			client.Lock.Unlock()
			continue

		case SekaiApiHttpStatusSessionError:
			mgr.Logger.Warnf("%s Server cookies expired, re-parsing...", strings.ToUpper(string(mgr.Server)))
			if err := mgr.parseCookies(ctx); err != nil {
				client.Lock.Unlock()
				resp := HarukiSekaiAPIFailedResponse{
					Result:  "failed",
					Status:  http.StatusForbidden,
					Message: fmt.Sprintf("Failed to parse cookies: %v", err),
				}
				return resp, http.StatusForbidden, err
			}
			retryCount++
			time.Sleep(retryDelay)
			client.Lock.Unlock()
			continue

		case SekaiApiHttpStatusUnderMaintenance:
			client.Lock.Unlock()
			resp := HarukiSekaiAPIFailedResponse{
				Result:  "failed",
				Status:  http.StatusServiceUnavailable,
				Message: fmt.Sprintf("%s Game server is under maintenance.", strings.ToUpper(string(mgr.Server))),
			}
			return resp, http.StatusServiceUnavailable, NewUnderMaintenanceError()

		case SekaiApiHttpStatusOk:
			result, err := client.handleResponse(*response)
			client.Lock.Unlock()
			if err != nil {
				resp := HarukiSekaiAPIFailedResponse{
					Result:  "failed",
					Status:  response.StatusCode(),
					Message: err.Error(),
				}
				return resp, response.StatusCode(), err
			}
			if m, ok := result.(map[string]any); ok {
				return m, response.StatusCode(), nil
			}
			return result, response.StatusCode(), nil

		default:
			client.Lock.Unlock()
			resp := HarukiSekaiAPIFailedResponse{
				Result:  "failed",
				Status:  response.StatusCode(),
				Message: fmt.Sprintf("Unexpected status code: %d", response.StatusCode()),
			}
			return resp, response.StatusCode(), fmt.Errorf("unexpected status code: %d", response.StatusCode())
		}
	}

	resp := HarukiSekaiAPIFailedResponse{
		Result:  "failed",
		Status:  http.StatusInternalServerError,
		Message: "Max retry attempts reached",
	}
	return resp, http.StatusInternalServerError, fmt.Errorf("max retry attempts reached")
}

func (mgr *SekaiClientManager) GetCPMySekaiImage(path string) ([]byte, error) {
	client := mgr.getClient()
	return client.GetCPMySekaiImage(path)
}

func (mgr *SekaiClientManager) GetNuverseMySekaiImage(userID, index string) ([]byte, error) {
	client := mgr.getClient()
	return client.GetNuverseMySekaiImage(userID, index)
}
